Adds feature to drop around a percentage

leandromoreira opened the pull request:
Originally from https://github.com/elasticsearch/logstash/pull/1073

leandromoreira committed:
Adds feature to drop around a percentage

Originally from https://github.com/elasticsearch/logstash/pull/1073

talevy commented on lib/logstash/filters/drop.rb:
```diff
@@ -26,6 +39,6 @@ def register
   public
   def filter(event)
-    event.cancel
+    event.cancel if rand < (@sample / 100.0)
```
to prevent the extra computation of `rand` at every drop, maybe we can 
check for whether sampling is even enabled with something like this and 
short circuit the computation if unnecessary

``` ruby
event.cancel if (@sample and rand < (@sample / 100.0))
```

jordansissel replied:
the `if @sample` requires the removal of `:default => ...` which is a good thing anyway :)

jordansissel commented on lib/logstash/filters/drop.rb:
```diff
@@ -18,7 +18,20 @@
+  #         } 
+  #       }
+  #     }
+  config :sample, :validate => :number, :default => 100
```
Can we call this `percentage` ? "sample => 50" sounds weird in my head when I read it.

Thoughts on an alternative name?

leandromoreira commented:
@talevy I liked your suggestion but since the property assumes a default value of 100, it would be better if we do this:

``` ruby
(@sample<100 and rand < (@sample / 100.0))
```

leandromoreira commented:
@jordansissel I vote + for percentage it makes the code easier to read.

leandromoreira committed:
Better performance and code readability 

Suggestion made at https://github.com/logstash-plugins/logstash-filter-drop/pull/2

leandromoreira commented:
@talevy I ran 

``` ruby
require 'benchmark'

@sample = 45

def sum(x)
  1 + 2 + x
end

puts "#" * 20
puts Benchmark.measure { 10000000.times { sum @sample } }
puts "#" * 20
puts Benchmark.measure { 10000000.times { sum(@sample) if (@sample < 100 and rand < (@sample / 100.0)) } }
puts "#" * 20
```

And the results (10000000.times 1.367734 vs 2.962667) were not so different, in this case I think leaving the code simpler would be better, what you think?

I ran similar bench before with old code and the results were similar ;/ https://github.com/elasticsearch/logstash/pull/1073#issuecomment-71288096

leandromoreira committed:
Reverting micro optimization 

Since it didn't prove to be faster https://github.com/logstash-plugins/logstash-filter-drop/pull/2#issuecomment-71289831

leandromoreira commented:
I "benchmarked" it wrongly :hankey:  the right way would be simulating `@sample =100`, and with this case (default) it performs better.
Sorry guys.

leandromoreira committed:
Improves performance for default case

The default case is dropping 100%

jsvd commented:
LGTM

wiibaa commented:
@jsvd with this feature where events would survive the drop filter, would it make sense to call filter_matched() on them to allow to add_tag and add_field ?

jsvd commented:
yes, yes it would.
